internal/util/merge/merge3.go

```diff
@@ -0,0 +1,165 @@
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/pathutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	mergeSourceAnnotation = "config.kubernetes.io/merge-source"
+	mergeSourceOriginal   = "original"
+	mergeSourceUpdated    = "updated"
+	mergeSourceDest       = "dest"
+)
+
+// Merge3 performs a 3-way merge on the original, upstream and
+// destination packages. It provides support for doing this only for
+// the parent package and ignore any subpackages. Whenever the boundaries
+// of a package differs between original, upstream and destination, the
+// boundaries in destination will be used.
+type Merge3 struct {
+	OriginalPath       string
+	UpdatedPath        string
+	DestPath           string
+	MatchFilesGlob     []string
+	MergeOnPath        bool
+	IncludeSubPackages bool
+}
+
+func (m Merge3) Merge() error {
+	// If subpackages are not included when doing the merge, first
+	// look up the known subpackages in destination so we can make sure
+	// those are ignored when reading files from original and updated.
+	var relPaths []string
+	if !m.IncludeSubPackages {
+		var err error
+		relPaths, err = m.findExclusions()
+		if err != nil {
+			return err
+		}
+	}
+
+	var inputs []kio.Reader
+	dest := &kio.LocalPackageReadWriter{
+		PackagePath:        m.DestPath,
+		MatchFilesGlob:     m.MatchFilesGlob,
+		SetAnnotations:     map[string]string{mergeSourceAnnotation: mergeSourceDest},
+		IncludeSubpackages: m.IncludeSubPackages,
+		PackageFileName:    kptfile.KptFileName,
+	}
+	inputs = append(inputs, dest)
+
+	// Read the original package
+	inputs = append(inputs, PruningLocalPackageReader{
+		LocalPackageReader: kio.LocalPackageReader{
+			PackagePath:        m.OriginalPath,
+			MatchFilesGlob:     m.MatchFilesGlob,
+			SetAnnotations:     map[string]string{mergeSourceAnnotation: mergeSourceOriginal},
+			IncludeSubpackages: m.IncludeSubPackages,
+			PackageFileName:    kptfile.KptFileName,
+		},
+		Exclusions: relPaths,
+	})
+
+	// Read the updated package
+	inputs = append(inputs, PruningLocalPackageReader{
+		LocalPackageReader: kio.LocalPackageReader{
+			PackagePath:        m.UpdatedPath,
+			MatchFilesGlob:     m.MatchFilesGlob,
+			SetAnnotations:     map[string]string{mergeSourceAnnotation: mergeSourceUpdated},
+			IncludeSubpackages: m.IncludeSubPackages,
+			PackageFileName:    kptfile.KptFileName,
+		},
+		Exclusions: relPaths,
+	})
+
+	kyamlMerge := filters.Merge3{
+		MergeOnPath: m.MergeOnPath,
+	}
+
+	return kio.Pipeline{
+		Inputs:  inputs,
+		Filters: []kio.Filter{kyamlMerge},
+		Outputs: []kio.Writer{dest},
+	}.Execute()
+}
+
+func (m Merge3) findExclusions() ([]string, error) {
+	var relPaths []string
+	paths, err := pathutil.DirsWithFile(m.DestPath, kptfile.KptFileName, true)
+	if err != nil {
+		return relPaths, err
+	}
+
+	for _, p := range paths {
+		rel, err := filepath.Rel(m.DestPath, p)
+		if err != nil {
+			return relPaths, err
+		}
+		if rel == "." {
+			continue
```

Comment:
Just to confirm that this step is to make sure we would never return the root package path correct ?